### PR TITLE
Feat - upgrade the cluster

### PR DESCRIPTION
**Pull request by chpl**
Continues the work done by @liranfarage89 in https://github.com/env0/k8s-modules/pull/32

### Breaking changes:
- there isn't upgrade path from previous version of this module
- only full blown installation is supported

### Feature Request
- Be able to install the eks cluster with latest version
- use as uptodate versions of the modules and providers as possible

### Solution
- update providers and modules
- some adjustments

### QA
- [x] deployed a cluster using the module

### CODE REVIEW

File: aws/autoscaler/main.tf [hunk 1]
```diff
@@ -1,40 +1,36 @@
...
+  # make sure that chart version matches the cluster version
+  helm_chart_version = "9.33.0"
 
   values = yamlencode({
-    # Determined by our cluster version -
-    # Open the Cluster Autoscaler releases page from GitHub in a web browser and find the latest Cluster Autoscaler version that matches the Kubernetes major and minor version of your cluster.
-    # For example, if the Kubernetes version of your cluster is 1.20, find the latest Cluster Autoscaler release that begins with 1.20.
-    "image" : {
-      "tag" : "v${data.aws_eks_cluster.cluster.version}.0"
-    }
```
Comment (chpl) [inline aws/autoscaler/main.tf:37]: the image should be in sync with the chart version, so it's better not to override it here.

File: aws/csi-driver/main.tf [hunk 2]
```diff
@@ -1,140 +1,77 @@
...
-            "${local.oidc_without_http}:sub" : "system:serviceaccount:kube-system:efs-csi-controller-sa"
-          }
-        }
-      }
-    ]
-  })
-  inline_policy {
-    name = "${var.cluster_name}_AmazonEKS_EFS_CSI_Driver_Policy"
+module "efs_csi_role" {
+  source  = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"
```
Comment (chpl) [inline aws/csi-driver/main.tf:9]: use module to create the role

File: aws/csi-driver/main.tf [hunk 3]
```diff
@@ -1,140 +1,77 @@
...
-            }
+  create_role = true
+  role_name   = local.role_name
 
-          }
-        }
-      ]
-    })
+  role_policy_arns = {
+    "AmazonEKS_CSI_EFS_Policy" = "arn:aws:iam::aws:policy/service-role/AmazonEFSCSIDriverPolicy"
```
Comment (chpl) [inline aws/csi-driver/main.tf:16]: use AWS managed policy

File: aws/csi-driver/main.tf [hunk 4]
```diff
@@ -1,140 +1,77 @@
...
-    annotations = {
-      "meta.helm.sh/release-name": "aws-efs-csi-driver"
-      "meta.helm.sh/release-namespace": "kube-system"
-      "eks.amazonaws.com/role-arn": aws_iam_role.role_with_web_identity_oidc.arn
+  oidc_providers = {
+    external_dns = {
+      provider_arn = var.oidc_provider_arn
+      namespace_service_accounts = [
+        "${local.namespace}:${local.controller_service_account_name}",
+        "${local.namespace}:${local.node_service_account_name}"
```
Comment (chpl) [inline aws/csi-driver/main.tf:24]: two SAs

File: aws/csi-driver/main.tf [hunk 5]
```diff
@@ -1,140 +1,77 @@
...
-          }
-        }
-      ]
-    })
+  role_policy_arns = {
+    "AmazonEKS_CSI_EFS_Policy" = "arn:aws:iam::aws:policy/service-role/AmazonEFSCSIDriverPolicy"
   }
-}
 
-resource "kubernetes_service_account" "csi_service_account" {
```
Comment (chpl) [inline aws/csi-driver/main.tf:78]: let the chart create it

File: aws/csi-driver/main.tf [hunk 6]
```diff
@@ -1,140 +1,77 @@
...
+      }
+      node = {
+        serviceAccount = {
+          create = true
+          name   = local.node_service_account_name
+          annotations = {
+            "eks.amazonaws.com/role-arn" = module.efs_csi_role.iam_role_arn
+          }
+        }
+      }
```
Comment (chpl) [inline aws/csi-driver/main.tf:58]: let the chart create service accounts

File: aws/csi-driver/versions.tf [hunk 7]
```diff
@@ -1,13 +1,9 @@
...
+  required_version = ">= 1.3.0"
 
   required_providers {
-    aws        = "~> 3.68.0"
-    kubernetes = "~> 2.7.0"
-    helm       = "~> 2.4.0"
-    kubectl = {
-      source  = "gavinbunney/kubectl"
-      version = ">= 1.7.0"
-    }
```
Comment (chpl) [inline aws/csi-driver/versions.tf:11]: no need for this one

File: aws/csi-driver/main.tf (hunk 6 above)
Comment (chpl) [inline aws/csi-driver/main.tf:58]: let the chart create service accounts

File: aws/eks/main.tf [hunk 8]
```diff
@@ -1,46 +1,73 @@
...
-  version         = "17.24.0"
+  source  = "terraform-aws-modules/eks/aws"
+  version = "~> 20.8"
+
   cluster_name    = var.cluster_name
-  cluster_version = "1.23"
-  subnets         = data.aws_subnet_ids.private.ids
-  enable_irsa     = true
-  vpc_id          = var.vpc_id
+  cluster_version = var.kubernetes_version
```
Comment (chpl) [inline aws/eks/main.tf:10]: get it from var

File: aws/eks/main.tf [hunk 9]
```diff
@@ -1,46 +1,73 @@
...
+
+  enable_irsa = true
+
+  enable_cluster_creator_admin_permissions = true
 
-  cluster_enabled_log_types = ["api", "scheduler", "controllerManager"] # https://docs.aws.amazon.com/eks/latest/userguide/control-plane-logs.html
+  access_entries = var.cluster_access_entries
```
Comment (chpl) [inline aws/eks/main.tf:16]: new way to configure cluster access

File: aws/eks/main.tf [hunk 10]
```diff
@@ -1,46 +1,73 @@
...
+
+  cluster_endpoint_public_access = true
+
+  eks_managed_node_group_defaults = {
     ami_type  = "AL2_x86_64"
     disk_size = 50
   }
 
-  workers_group_defaults = {
-    protect_from_scale_in = true
```
Comment (chpl) [inline aws/eks/main.tf:26]: not sure why this was enabled.
[docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-instance-protection.html](https://docs.aws.amazon.com/autoscaling/ec2/userguide/ec2-auto-scaling-instance-protection.html)

File: aws/eks/main.tf [hunk 11]
```diff
@@ -1,46 +1,73 @@
...
+      most_recent = true
+    }
+    vpc-cni = {
+      most_recent = true
+    }
+  }
 
-  map_roles    = var.map_roles
-  map_users    = []
-  map_accounts = []
```
Comment (chpl) [inline aws/eks/main.tf:45]: see `access_entries` above

File: aws/eks/variables.tf [hunk 12]
```diff
@@ -1,30 +1,23 @@
...
 variable "instance_type" {
   default = "t3a.2xlarge" # 8vCPUs 32GB
 }
 
-variable "write_kubeconfig" {
-  type = bool
-  default = false 
-}
\ No newline at end of file
+variable "cluster_access_entries" {}
```
Comment (chpl) [inline aws/eks/variables.tf:23]: new way to manage cluster access - replaces `map_roles`

File: aws/vpc/main.tf [hunk 13]
```diff
@@ -1,12 +1,12 @@
...
   source  = "terraform-aws-modules/vpc/aws"
-  version = "3.11.0"
+  version = "~> 5.8"
 
   name                 = "vpc-${var.cluster_name}"
   cidr                 = var.cidr
-  azs                  = data.aws_ec2_instance_type_offerings.supported_azs.locations
-  private_subnets      = var.private_subnets
-  public_subnets       = var.public_subnets
+  azs                  = var.azs
```
Comment (chpl) [inline aws/vpc/main.tf:7]: Pass the AZs as var - 
autodetection feature can cause the VPC to be recreated - not something we want

**Comment by chpl** [comment]
[Make `Kushield Cluster` deployment work again](https://trello.com/c/P1ZN3oX8/11135-make-kushield-cluster-deployment-work-again)

**Review by liranfarage89 — APPROVED** [review]
LGTM, I didn't see anything I can comment about. See the merge conflicts

**Comment by chpl** [comment]
@away168 FYI - as stated in the description there isn't upgrade path from previous version of this module, so I want to tag the master as `v0.4.0` and after merging this it will be `v1.0.0`

**Pull request merged**